File: Source/WickedSick.KineticGraph.Controls/Physics/ForceHelper.cs
﻿using System;
using System.Windows;

namespace WickedSick.KineticGraph.Controls.Physics
{
    internal static class ForceHelper
    {
        public const double DEFAULT_ATTRACTION_CONSTANT = 0.00000004;
        private const int MAGNITUDE_MAX = 10;

        static ForceHelper()
        {
            AttractionConstant = DEFAULT_ATTRACTION_CONSTANT;
        }

        public static double AttractionConstant { get; set; }

        /// <summary>
        /// Applies coulomb repulsion to both points
        /// </summary>
        /// <param name="a"></param>
        /// <param name="b"></param>
        /// <param name="k">coefficient of attraction</param>
        /// <param name="f1">Force at point a</param>
        /// <param name="f2">Force at point b</param>
        public static void ApplyCoulombRepulsion(NodeState a, NodeState b, double k)
        {
            double dx = a.Position.X - b.Position.X;
            double dy = a.Position.Y - b.Position.Y;
            double sqDist = dx * dx + dy * dy;
            if (sqDist == 0)
                return;
            double d = Math.Sqrt(sqDist);

            double mag = 1.0 / sqDist; // Force magnitude

            mag -= AttractionConstant * d; // plus WEAK attraction

            mag *= k;

            if (mag > MAGNITUDE_MAX)
                mag = MAGNITUDE_MAX; // Clip maximum

            double tempX = mag * (dx / d);
            double tempY = mag * (dy / d);

            if (!a.IsFrozen)
            {
                a.Force.X += tempX;
                a.Force.Y += tempY;
            }

            if (!b.IsFrozen)
            {
                b.Force.X -= tempX;
                b.Force.Y -= tempY;
            }
        }

        public static void ApplyHookeAttraction(Point a, Point b, double k, ref Point f)
        {
            //double dx = (a.X - b.X);
            //double dy = (a.Y - b.Y);
            //double d = Math.Sqrt(dx * dx + dy * dy);

            //double mag = -k * 0.001 * Math.Pow(d, 1.0); // Force magnitude
            //double mag = -k * 0.001 * d; // Force magnitude

            f.X += -k * (a.X - b.X);
            f.Y += -k * (a.Y - b.Y);
        }

        public static void ApplyHookeAttraction(NodeState a, NodeState b, double k)
        {
            var p1 = a.Position;
            var p2 = b.Position;
            
            var x = -k * (p1.X - p2.X);
            var y = -k * (p1.Y - p2.Y);

            if (!a.IsFrozen)
            {
                a.Force.X += x;
                a.Force.Y += y;
            }

            if (!b.IsFrozen)
            {
                b.Force.X -= x;
                b.Force.Y -= y;
            }
        }
    }
}

File: Source/WickedSick.KineticGraph.Controls.Fayde/Physics/ForceHelper.ts
module KineticGraph.Controls.Physics {
    export interface IVector {
        X: number;
        Y: number;
    }
    export interface INodeState {
        IsFrozen: boolean;
        Position: IVector;
        Velocity: IVector;
        Force: IVector;
    }

    var DEFAULT_ATTRACTION_CONSTANT = 0.000000004;
    var MAGNITUDE_MAX = 10;
    export class ForceHelper {
        static AttractionConstant: number = DEFAULT_ATTRACTION_CONSTANT;
        /// Applies coulomb repulsion to both points
        static ApplyCoulombRepulsion(a: INodeState, b: INodeState, k: number) {
            var dx = a.Position.X - b.Position.X;
            var dy = a.Position.Y - b.Position.Y;
            var sqDist = dx * dx + dy * dy;
            if (sqDist == 0)
                return;
            var d = Math.sqrt(sqDist);

            var mag = 1.0 / sqDist; // Force magnitude

            mag -= ForceHelper.AttractionConstant * d; // plus WEAK attraction

            mag *= k;

            if (mag > MAGNITUDE_MAX)
                mag = MAGNITUDE_MAX; // Clip maximum

            var tempX = mag * (dx / d);
            var tempY = mag * (dy / d);

            if (!a.IsFrozen) {
                a.Force.X += tempX;
                a.Force.Y += tempY;
            }

            if (!b.IsFrozen) {
                b.Force.X -= tempX;
                b.Force.Y -= tempY;
            }
        }
        static ApplyHookeAttraction(a: INodeState, b: INodeState, k: number) {
            var p1 = a.Position;
            var p2 = b.Position;

            var x = -k * (p1.X - p2.X);
            var y = -k * (p1.Y - p2.Y);

            if (!a.IsFrozen) {
                a.Force.X += x;
                a.Force.Y += y;
            }

            if (!b.IsFrozen) {
                b.Force.X -= x;
                b.Force.Y -= y;
            }
        }
    }
}